### ISSUE TITLE
Detect and aggregate parcel data within the OA ingestion process

### ISSUE THREAD

**Issue by waldoj** [openaddresses/openaddresses#1587]
A fair number of OpenAddress' data sources include parcel data, but we're not doing anything with this data. We just extract the coordinates and the address. We should put parcel data to work to work, providing output in the form of open parcel data in parallel to open address data.

What this probably looks like is examining [the state.txt output](http://results.openaddresses.io/state.txt) to find polygon data, retrieve and aggregate those records, and see what that starting point looks like. Then figure out a path to modifying the JSON manifests for each locality to include parcels (which are often found at a different source entirely), processing those input sources, and aggregating them, all in the manner of and in concert with existing OA infrastructure.

**Comment by NelsonMinar** [comment]
As for implementation, I'm in favor of a system where the Python machine code also emits polygons when processing source input. At a high level, the idea is that the same code that downloads sources and emits address CSV can also emit parcel shapefiles. At a low level, I'd start by looking at [ogr_source_to_csv](https://github.com/openaddresses/machine/blob/master/openaddr/conform.py#L511) and considering whether a similar function could emit polygons as an optional extra step in the conform pipeline.

For input, the scripts will need something like the existing conform JSON that describes how to get the parcels from any particular source. That could be an extra set of fields in the same JSON file or a second sidecar JSON file; I suspect maintaining a single file is better.

For outputs, I would love to maintain the simplicity of "one source is one CSV output" that we have now. Only now we have a second output, which is a file containing geometry. I'm agnostic about the format of the output: shapefile, GeoJSON, or WKT in a CSV column. (Note that really each source produces several output files which we then zip together, so the parcels would just be more.)

One open question in my mind: are there important metadata properties for parcels we should capture? Surely we don't just want the geometry, but also the address name, etc. If so, what is the schema for that metadata? Is there one single schema we can convert all sources to, like we do for address records? Do we need something as complex as our existing conform code's specification language for extracting and transforming the metadata?

**Comment by migurski** [comment]
I think we’re covered on polygons included in the Machine code. They can be found in the cached data for each source!

For input, I have a weak preference for using the existing source files and extending them with additional columns. We can expand the supported attributes if need be.

**Comment by stoneyv** [comment]
I examined the cache files for some entries in the state.txt ouput that was referenced by @waldoj  I found lat, long coordinates for address parcels, but I did not see any polygon data.  I looked at the .shp file with qgis and a head of the import for .dbf files
ar/ba/buenos_aires.json
us/ca/alameda.json
us/or/portland_metro.json

**Comment by migurski** [comment]
Check out the `geometry type` header in state.txt, and look at us/ca/berkeley: it has polygons.

**Comment by stoneyv** [comment]
OK, state.txt entries with field 'geometry type' equals Polygon have parcel polygon data.  I happened to pick examples that had a type of 'Point'.

**Comment by stoneyv** [comment]
I was reading the contributing document to learn how you should extend the existing source file.  How would you add the additional column(s)?  master/sources/us/ca/berkeley.json

**Comment by migurski** [comment]
Here's a record of the last time we added attributes: https://github.com/openaddresses/openaddresses-ops/issues/7

At that time, we needed to modify the machine code as we went in order to use the new attributes. In this case, it might not be necessary to synchronize across multiple repositories.

**Comment by waldoj** [comment]
I just talked to a vendor, who had these two utterly reasonable answers for which I do not have canonical answers:

**1. How easy is this to clone and stand up a development instance of?** We don't have a DEVELOPERS.md or anything like that, at least that I've ever seen.

**2. What does the deliverable look like?** Like, specifically. I've been thinking in terms of processes and big-picture outcomes, but of course a developer has to know what thing it is that they are creating.

I've got my own ideas about #2, but I'd like us to discuss that here, to reach consensus.

**Comment by migurski** [comment]
My preference would be not ask someone to modify the code in the machine repository. It’s complex and tightly optimized for producing exactly one kind of output; I imagine the scope would easily double or triple in size and uncertainty if we asked a developer to understand too. Rather, I think it would be better to use http://results.openaddresses.io/index.json as a starting point, and treat the Parcels process as a downstream product from the address point process.

**Comment by waldoj** [comment]
Here's a first crack at defining the abilities of the deliverable:

* Identify which sources contain parcel data.
* Identify the file format.
* Convert the contents into CSV, including the UID, boundaries, and address (as extant).

I _think_ this is the minimum viable product, although I imagine there are a bunch of existing standards and practices to adhere to that I don't know the first thing about.

**Comment by NelsonMinar** [comment]
That sounds good @waldoj. I'd break the parcel system up into two parts:

1. A standalone command line program that can take a source file as input, and produce parcel CSV as output. The equivalent of `openaddr-process-one` but for parcels.
2. An automated system for running all parcels regularly and/or in response to checkins.

@migurski, what do you think about #2? I'm really hoping we can re-use your existing CI and batch management code, because reimplementing that seems insane. If #1 is done as an external downstream product, how does the larger job management work?

I've been saying all along that I'd do #1 as a modification to the existing code. I think it'd be relatively simple, but I defer to @migurksi's judgement that it wouldn't be.

**Comment by migurski** [comment]
I’m thinking we can implement it in steps; that's how I’d do it. Start with a manual script that requires a human to push a button, move on to a weekly(ish) scheduled task, and then finally integrate with the CI system so it's event-driven with everything else.

**Comment by NelsonMinar** [comment]
Agreed that rolling out in steps will work. I just think part of the deliverable should be that it integrates with the existing machinery, or at least that we can then integrate it ourselves.

**Comment by andrewharvey** [comment]
hopefully this will be possible with https://github.com/openaddresses/openaddresses-ops/issues/21
